File: lecture_ws/src/lecture_pkg/scripts/simple_action_server.py
#! /usr/bin/env python

import rospy
import actionlib
from lecture_pkg.msg import *

class DoDishesServer:
    _feedback = lecture_pkg.msg.DoDishesFeedback()
    _result = lecture_pkg.msg.DoDishesResult()
    
    def __init__(self):
        self.server = actionlib.SimpleActionServer('do_dishes', DoDishesAction,self.execute,False)
        self.server.start()

    def execute(self, goal):
        print "Requesting dishwasher %d"%(goal.dishwasher_id)
        result = self.server.get_default_result()
        result.total_dishes_cleaned = 100
        for i in range(1,10):
            self._feedback.percent_complete = 10*i
            self.server.publish_feedback(self._feedback)
        print "Returning dishes_cleaned %d"%(result.total_dishes_cleaned)
        self.server.set_succeeded(result)

if __name__ == '__main__':
    rospy.init_node('do_dishes_server')
    server = DoDishesServer()
    rospy.spin()

File: lecture_ws/src/lecture_pkg/src/simple_action_server.cpp
#include <lecture_pkg/DoDishesAction.h>
#include <actionlib/server/simple_action_server.h>

typedef actionlib::SimpleActionServer<lecture_pkg::DoDishesAction> Server;

void execute(const lecture_pkg::DoDishesGoalConstPtr& goal, Server* as)
{
  lecture_pkg::DoDishesFeedback feedback_;
  lecture_pkg::DoDishesResult result_;
  for (int i = 0; i <= 10; i++)
  {
    feedback_.percent_complete = 10 * i;
    as->publishFeedback(feedback_);
  }
  as->setSucceeded();
}

int main (int argc, char** argv)
{
  ros::init(argc, argv, "do_dishes_server");
  ros::NodeHandle n;
  Server server(n, "do_dishes", boost::bind(&execute, _1, &server), false);
  server.start();
  ros::spin();
  return 0;
}
